File: Chapter_3_Algorithms/3.1_Algorithms/Exercises/repo/source/solution_15.cpp
/*
  Discrete Mathematics and Its Applications by (Kenneth H. Rosen)
  Chapter 3.1
  Solution for problem 15
  Author: Rakesh Kumar (cpp.rakesh(at)gmail.com)
  Date: Jan 19th, 2017
 */

#include "common.h"
#include "merge_sort.h"
#include <cstdio>
#include <vector>

// Try to solve this problem with divide and conquer
// We will find a place where it suits exactly
int insert_pos(const std::vector<int>& list, const int x,  int start, int end) {
    if (start == end) {
        if (x < list[start] && start != 0)
            return start - 1;
        else
            return start;
    } else {
        const int mid = (start + end) >> 1;
        if (list[mid] == x)
            return mid;
        else if (x > list[mid] && x < list[mid + 1])
            return mid;
        else if (list[mid] > x)
            insert_pos(list, x, start, mid - 1);
        else
            insert_pos(list, x, mid + 1, end);
    }
}

void test_insert_pos() {
    discrete_mathematics::chapter_3::Common<int> common;
    discrete_mathematics::chapter_3::MergeSort<int> sort;

    for (int i = 0; i < 10; ++i) {
        printf("---------------------------------------------------------------------------------------------------------------\n");
        std::vector<int> list = common.build_vector(50, 0, 100);
        const int x = common.random(0, 100);
        sort.sort(list);
        common.print(list);
        printf("Insert position of [%3d] == [%3d]\n", x, insert_pos(list, x, 0, list.size() - 1));
        printf("---------------------------------------------------------------------------------------------------------------\n");        
    }
}

int main() {
    test_insert_pos();

    return 0;
}

File: chapter_9_relations/9.3_representing_relations/repo/solution_15.py
'''
    Problem 15
    Chapter 9.3
    Discrete Mathematics and its applications
    Rakesh Kumar
'''
def boolean_pow(A, B):
    R = []
    for i in range(len(A)):
        r = []
        for j in range(len(A)):
            v = 0
            for k in range(len(A)):
                if A[i][k] == 1 and B[k][j] == 1:
                    v = 1
            r.append(v)
        R.append(r)
    return R

def print_matrix(A):
    for r in range(len(A)):
        print('|', end='')
        for c in range(len(A[r])):
            print(f'{A[r][c]} ', end='')
        print('|')

if __name__ == '__main__':
    R = [[0,1,0], [0,0,1], [1,1,0]]
    print('a)')    
    A = boolean_pow(R, R)
    print_matrix(A)
    print('b)')
    B = boolean_pow(R, A)
    print_matrix(B)
    print('c)')
    C = boolean_pow(R, B)
    print_matrix(C)
    